### PR TITLE
proposal: ignore interfaces that contains type constraints

### PR DESCRIPTION
**Pull request by hoonmin**
This is a proposal PR to generate mocks without error for following cases:
```go
type Water[R any, C UnsignedInteger] interface {
	Fish(R) []C
}

type UnsignedInteger interface {
	~uint | ~uint32 | ~uint64
}
```

Go `types` package seems to wrap interfaces that contain type constraints by checking a type set literal of the form `~T` and `A|B`.
https://github.com/golang/go/blob/master/src/go/types/decl.go#L668

So gomock can just ignore that pattern to generate mocks safely without `don't know how to mock method of type` errors. I think this can be a better solution for custom type constraints than [`-exclude` flag](https://github.com/uber-go/mock/pull/72).

### CODE REVIEW

File: mockgen/generic_go118.go [hunk 1]
```diff
@@ -114,3 +126,5 @@ func (p *fileParser) parseGenericMethod(field *ast.Field, it *namedInterface, if
 
 	return p.parseMethod(nf, it, iface, pkg, tps)
 }
+
+var errIgnore = errors.New("parser decided to ignore the interface for constraints")
```
Comment (JacobOaks) [inline mockgen/generic_go118.go:130]: Consider giving this a name that will make it clear what case we are ignoring in `parse.go`. We can also simplify the error message itself.

```suggestion
var errConstraintInterface = errors.New("interface contains constraints")
```

File: mockgen/parse.go [hunk 2]
```diff
@@ -232,6 +232,9 @@ func (p *fileParser) parseFile(importPath string, file *ast.File) (*model.Packag
 			continue
 		}
 		i, err := p.parseInterface(ni.name.String(), importPath, ni)
+		if err == errIgnore {
```
Comment (JacobOaks) [inline mockgen/parse.go:235]:
```suggestion
		if errors.Is(err, errConstraintInterface) {
```

This is a more safe check that will not break if, for some reason in a future update, the error gets wrapped.

**Review by JacobOaks — COMMENTED** [review]
I think this change is good - just a couple comments. Thanks!

**Comment by hoonmin** [comment]
> I think this change is good - just a couple comments. Thanks!

I learned some basic but important things from your review. I've applied all of your suggestions, thank you!

**Review by sywhang — APPROVED** [review]
This looks good to me. Thanks for making this change!

**Pull request merged**